src/coreclr/jit/importer.cpp

```diff
@@ -9839,11 +9841,51 @@ void Compiler::impImportBlockCode(BasicBlock* block)
 #endif
                 {
                     /* Create a call to 'new' */
+                    helper = info.compCompHnd->getNewArrHelper(resolvedToken.hClass);
 
                     // Note that this only works for shared generic code because the same helper is used for all
                     // reference array types
-                    op1 =
-                        gtNewHelperCallNode(info.compCompHnd->getNewArrHelper(resolvedToken.hClass), TYP_REF, op1, op2);
+                    op1 = gtNewHelperCallNode(helper, TYP_REF, op1, op2);
+                }
+
+                // If we're jitting a static constructor and detect the following code pattern:
+                //
+                //  newarr
+                //  stsfld
+                //  ret
+                //
+                // We replace default heap allocator for newarr with the one that prefers frozen segments.
+                // This is a very simple and conservative implementation targeting Array.Empty<T>(), ideally
+                // we want to be able to use frozen allocators more broadly, but that analysis is not trivial.
+                //
+                if (((info.compFlags & FLG_CCTOR) == FLG_CCTOR) &&
+                    // Does VM allow us to use frozen allocators? (e.g. are we in a non-collectible assembly)
+                    opts.jitFlags->IsSet(JitFlags::JIT_FLAG_FROZEN_ALLOC_ALLOWED))
+                {
+                    // Check next two opcodes
+                    const BYTE* nextOpcode1 = codeAddr + sizeof(mdToken);
+                    const BYTE* nextOpcode2 = nextOpcode1 + sizeof(mdToken) + 1;
+                    if (nextOpcode2 <= codeEndp && getU1LittleEndian(nextOpcode1) == CEE_STSFLD)
```

Comment:
What would it take to extend the pattern matcher to detect things like https://github.com/dotnet/runtime/blob/24bbee888f7b7e4b5c0c5af3448388f4edec218c/src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs#L34-L36 ?

Reply:
Yes, but I have a prototype for a more general phase that is able to track many things (e.g. delegates, large arrays, etc). The only problem that for that analysis I need optimizations to be enabled and we currently always compile CCTORs with Tier0 (although, we do apply optimizations for prejit) - I was thinking of "switch to optimized if current CCTOR doesn't have branches/EH and has alloctors"

Reply:
If you do that just be aware that some CCTORS are really large (especially those that initialize arrays), so the jit may decide not to optimize for other reasons.